geopolars/src/spatial_index.rs

```diff
@@ -0,0 +1,216 @@
+            [envelope.max().x, envelope.max().y],
+        ])
+    }
+}
+
+impl From<MultiPolygon<f64>> for NodeEnvelope {
+    fn from(multi_polygon: MultiPolygon<f64>) -> Self {
+        let envelope = multi_polygon.bounding_rect().unwrap();
+        NodeEnvelope::BBox([
+            [envelope.min().x, envelope.min().y],
+            [envelope.max().x, envelope.max().y],
+        ])
+    }
+}
+
+impl From<MultiPoint<f64>> for NodeEnvelope {
+    fn from(multi_point: MultiPoint<f64>) -> Self {
+        let envelope = multi_point.bounding_rect().unwrap();
+        NodeEnvelope::BBox([
+            [envelope.min().x, envelope.min().y],
+            [envelope.max().x, envelope.max().y],
+        ])
+    }
+}
+
+impl From<LineString<f64>> for NodeEnvelope {
+    fn from(line: LineString<f64>) -> Self {
+        let envelope = line.bounding_rect().unwrap();
+        NodeEnvelope::BBox([
+            [envelope.min().x, envelope.min().y],
+            [envelope.max().x, envelope.max().y],
+        ])
+    }
+}
+
+impl From<MultiLineString<f64>> for NodeEnvelope {
+    fn from(multi_line: MultiLineString<f64>) -> Self {
+        let envelope = multi_line.bounding_rect().unwrap();
+        NodeEnvelope::BBox([
+            [envelope.min().x, envelope.min().y],
+            [envelope.max().x, envelope.max().y],
+        ])
+    }
+}
+
+impl From<Line<f64>> for NodeEnvelope {
+    fn from(line: Line<f64>) -> Self {
+        let envelope = line.bounding_rect();
+        NodeEnvelope::BBox([
+            [envelope.min().x, envelope.min().y],
+            [envelope.max().x, envelope.max().y],
+        ])
+    }
+}
+
+pub struct TreeNode {
+    pub index: usize,
+    pub envelope: NodeEnvelope,
+}
+
+impl RTreeObject for TreeNode {
+    type Envelope = AABB<[f64; 2]>;
+
+    fn envelope(&self) -> Self::Envelope {
+        match self.envelope {
+            NodeEnvelope::Point(point) => AABB::from_point(point),
+            NodeEnvelope::BBox(bbox) => AABB::from_corners(bbox[0], bbox[1]),
+        }
+    }
+}
+
+impl TryFrom<Geometry<f64>> for NodeEnvelope {
+    type Error = PolarsError;
+    fn try_from(geom: Geometry<f64>) -> Result<Self, Self::Error> {
+        match geom {
+            Geometry::Polygon(poly) => Ok(poly.into()),
+            Geometry::MultiPolygon(multi_poly) => Ok(multi_poly.into()),
+            Geometry::Point(point) => Ok(point.into()),
+            Geometry::MultiPoint(multi_point) => Ok(multi_point.into()),
+            Geometry::Line(line) => Ok(line.into()),
+            Geometry::LineString(line_string) => Ok(line_string.into()),
+            Geometry::MultiLineString(multi_line_string) => Ok(multi_line_string.into()),
+
+            _ => Err(PolarsError::ComputeError(std::borrow::Cow::Borrowed(
+                "Geometry type not currently supported for indexing",
+            ))),
+        }
+    }
+}
+
+pub struct SpatialIndex {
+    pub r_tree: RTree<TreeNode>,
+}
+
+impl SpatialIndex {}
```

Comment:
Eventually going to turn this in to a Trait and implement it for SpatialIndex. Probably rename SpatialIndex to RTreeIndex or something similar